Add more supports for vsphere.

luomiao opened the pull request:
Accept vSphere's server, datacenter, cluster setting by flags
"vsphere-server", "vsphere-datacenter", and "vsphere-resource-pool".
Username and password can be set by environment variables:
"VSPHERE_USERNAME" and "VSPHERE_PASSWORD".

Maybe we will have to give all the vCenter related information in environment variable way.
But now let's use flags for easy testing.

Commit:
Add more supports for vsphere.

Accept vSphere's server, datacenter, cluster setting by flags
"vsphere-server", "vsphere-datacenter", and "vsphere-resource-pool".
Username and password can be set by environment variables:
"VSPHERE_USERNAME" and "VSPHERE_PASSWORD".

prashima commented on upup/pkg/fi/cloudup/apply_cluster.go:
```diff
@@ -365,11 +365,8 @@ func (c *ApplyClusterCmd) Run() error {
-			//	return fmt.Errorf("GCE support is currently alpha, and is feature-gated.  export KOPS_FEATURE_FLAGS=AlphaAllowGCE")
-			//}
+			// TODO: map region with vCenter cluster, or datacenter, or datastore?
+			region = vsphereCloud.Cluster
```
Not specific to this change, but we need to add a feature flag for our vSphere.

luomiao replied:
Feature flag here means?

SandeepPissay replied:
As long as all our code is enabled only with vsphere cloud provider flag then I dont see a need for feature flag.

prashima replied:
GCE also has a feature flag https://github.com/kubernetes/kops/blob/f53070455051dca489632b9c47bc4ac6333e5c2f/upup/pkg/fi/cloudup/apply_cluster.go#L57. I think it makes sense to prevent use of vSphere as cloud provider till we are ready for people to try it out. Let's discuss this separately though.

prashima commented on upup/pkg/fi/cloudup/vsphere/vsphere_cloud.go:
```diff
@@ -36,6 +40,18 @@ func (c *VSphereCloud) ProviderID() fi.CloudProviderID {
+	cluster := spec.VSphereResourcePool
+	username := os.Getenv("VSPHERE_USERNAME")
+	password := os.Getenv("VSPHERE_PASSWORD")
+
```
What will happen if the environment variable is not set? At which point will the user start to see the problem? 
Will it be better to check these env variables for emptiness in create_cluster.go?

luomiao replied:
Not a problem for now, since we don't really create a client yet. But if the flags are not set, there will be an error jumping out earlier. Now it could be just some dummy values.

prashima replied:
Ok, let's make sure that when we make use of these values, user does get relevant message.

luomiao replied:
oh sorry I think I read your original comments wrong :)
Yes you are right we should check the emptiness.
Let me add the check.

prashima commented on upup/pkg/kutil/delete_cluster_vsphere.go:
```diff
@@ -0,0 +1,23 @@
+	"fmt"
+)
+
+func (c *AwsCluster) listResourcesVSphere() (map[string]*ResourceTracker, error) {
```
Not specific to your change- AwsCluster should be renamed to something generic. Let's wait for kops master branch to have this change. Else we can do it before creating a pull request for kops master.

luomiao replied:
Yeah we can have it this way and fix it when they update the upstream. We will see the conflicts when we merge if they fixed it by that time.

prashima commented:
LGTM

SandeepPissay commented on cmd/kops/create_cluster.go:
```diff
@@ -181,6 +186,10 @@ func NewCmdCreateCluster(f *util.Factory, out io.Writer) *cobra.Command {
+	// vSphere flags
+	cmd.Flags().StringVar(&options.VSphereServer, "vsphere-server", options.VSphereServer, "vsphere-server is required for vSphere. Set vCenter URL Ex: 10.192.10.30 or myvcenter.io (without https://)")
+	cmd.Flags().StringVar(&options.VSphereDatacenter, "vsphere-datacenter", options.VSphereDatacenter, "vsphere-datacenter is required for vSphere. Set the name of the datacenter in which to deploy Kubernetes VMs.")
+	cmd.Flags().StringVar(&options.VSphereResourcePool, "vsphere-resource-pool", options.VSphereDatacenter, "vsphere-resource-pool is required for vSphere. Set a valid Cluster, Host or Resource Pool in which to deploy Kubernetes VMs.")
```
Why not just rely on environment variables for all vSphere things now? I dont see the benefit of these flags. I think we should either use all flags or all environment variables.

prashima replied:
I think it's good to have only the credentials as env variable. Reason being, flags and spec fields can be logged and will help in debugging. Users will end up setting too many env variables if all of these flags are converted to env variables.

luomiao replied:
I am on the flag side for similar reason...
It's hard to track env variables.
If the users don't set flags correctly, kops will stop at the beginning and return the suggestions like what we added here.
And these are only a few of them, later there might be even more, which makes env variable way even harder to use.

SandeepPissay replied:
> Reason being, flags and spec fields can be logged and will help in debugging

We can log environment variables too.

> Users will end up setting too many env variables if all of these flags are converted to env variables.

Its the same argument for flags too. So, I dont see how one is superior over another in this respect.

Regarding Miao's comments, I dont see how flags are better. Kops can still stop by validating the environment variables at the beginning. If we have more then having flags is as complex/simple as environment variables.

I dont mind using flags here but I just dont see how it s better over environment variables. In fact from security perspective, it is better to have them as environment variables as nobody else expect the one using kops will know the vsphere related parameters as they dont appear in the OS process list.

luomiao replied:
One reason for flags:
you may create two different clusters with two different vCenter.
And you may update them later separately. Using environment variables, they won't be saved into cluster.spec. And it's very easy to mess things up. Using flags, these information can be stored inside cluster.config, and you don't need to worry about the changing all the variables everytime when you switch to another cluster.

SandeepPissay commented on pkg/apis/kops/cluster.go:
```diff
@@ -241,6 +241,11 @@ type ClusterSpec struct {
 	CloudLabels map[string]string `json:"cloudLabels,omitempty"`
+
+	// vSphere required specs
+	VSphereServer       string
```
This struct does not seem to have anything cloud specific. This is cluster specific struct. Not sure if this is the right place for VC specific things.

luomiao replied:
The CloudLabels above these seems is a AWS specific things... So I suppose it's ok to add VC specific things?

luomiao replied:
OK seems cloudconfig is the right place. Let me move these fields to there.

Commit:
Address comments.

1. Check the emptiness of VSPHERE_USERNAME and VSPHERE_PASSWORD
2. Move vSphere specific fields from clusterSpec to clusterSpec.CloudConfig

luomiao commented:
Have addressed comments:
    1. Check the emptiness of VSPHERE_USERNAME and VSPHERE_PASSWORD
    2. Move vSphere specific fields from clusterSpec to clusterSpec.CloudConfig

SandeepPissay commented on cmd/kops/create_cluster.go:
```diff
@@ -472,6 +481,27 @@ func RunCreateCluster(f *util.Factory, out io.Writer, c *CreateClusterOptions) e
+			if c.VSphereServer == "" {
+				return fmt.Errorf("vsphere-server is required for vSphere. Set vCenter URL Ex: 10.192.10.30 or myvcenter.io (without https://)")
+			}
+			cluster.Spec.CloudConfig.VSphereServer = c.VSphereServer
```
Why should a k8s cluster spec have vSphere server details? Since kops will be used on a particular VC, storing this info here seems redundant. Can you explain what is the purpose this is serving?

luomiao replied:
You need spec.cloudconfig to include vSphere information. They will be needed for creating the cloud.config file for cloud provider later. GCE also have their cloudConfig info at the same place. Although they are not accepting user's own configuration, which we need.

SandeepPissay commented on upup/pkg/fi/cloudup/vsphere/vsphere_cloud.go:
```diff
@@ -36,6 +40,21 @@ func (c *VSphereCloud) ProviderID() fi.CloudProviderID {
 }
 
+func NewVSphereCloud(spec *kops.ClusterSpec) (*VSphereCloud, error) {
+	server := spec.CloudConfig.VSphereServer
```
I would suggest to convert all these flags to env variables. If not, atleast keep the VC IP as env variable.
